taipy/core/_orchestrator/_dispatcher/_job_dispatcher.py

```diff
@@ -55,30 +55,37 @@ def stop(self, wait: bool = True, timeout: Optional[float] = None):
             wait (bool): If True, the method will wait for the dispatcher to stop.
             timeout (Optional[float]): The maximum time to wait. If None, the method will wait indefinitely.
         """
-        self.stop_wait = wait
-        self.stop_timeout = timeout
         self._STOP_FLAG = True
+        if wait and self.is_alive():
+            self._logger.debug("Waiting for the dispatcher thread to stop...")
+            self.join(timeout=timeout)
 
     def run(self):
         self._logger.debug("Job dispatcher started.")
         while not self._STOP_FLAG:
-            try:
-                if self._can_execute():
-                    with self.lock:
-                        if self._STOP_FLAG:
-                            break
+            if not self._can_execute():
+                time.sleep(0.1)  # We need to sleep to avoid busy waiting.
+                continue
+
+            with self.lock:
+                self._logger.debug("Acquiring lock to check jobs to run.")
+                job = None
+                try:
+                    if not self._STOP_FLAG:
                         job = self.orchestrator.jobs_to_run.get(block=True, timeout=0.1)
-                    self._execute_job(job)
-                else:
-                    time.sleep(0.1)  # We need to sleep to avoid busy waiting.
-            except Empty:  # In case the last job of the queue has been removed.
-                pass
-            except Exception as e:
-                self._logger.exception(e)
-                pass
-        if self.stop_wait:
-            self._logger.debug("Waiting for the dispatcher thread to stop...")
-            self.join(timeout=self.stop_timeout)
+                except Empty:  # In case the last job of the queue has been removed.
+                    pass
+            self._logger.debug(f"Releasing lock after checking jobs to run.")
+            if job:
+                self._logger.debug(f"Got a job to execute {job.id}.")
+                try:
+                    if not self._STOP_FLAG:
+                        self._execute_job(job)
+                    else:
+                        self.orchestrator.jobs_to_run.put(job)
+                except Exception as e:
+                    self._logger.exception(e)
+        time.sleep(1)
```

Comment:
We sleep 1 second when stopping the dispatcher. The purpose is to stay alive and wait for the thread calling the stop method to have enough time to enter the if block (`if wait and self.is_alive():`), so we are sure the subthreads are joined before killing the thread.

Reply:
Ain't joined and killed thread the same thing?
The `thread.join()` will wait for the thread to stop anyway, and this `sleep()` makes it wait 1 second longer.

Reply:
In the stop method, there is an if condition to check if the dispatcher thread is running. If it is not running, we don't join the callbacks sub-threads. 

So there is a possibility to be be in the following situation:
- the STOP_FLAG is set to True in the stop method
- The dispatcher thread terminates by completing the main loop.
- We check if it is running to join the sub-threads but can't join them since the dispatcher thread is not running anymore.
- We terminate the tests by removing the entities
- The sub-threads keep running and try to access the entities

Does it make sense?

This sleep(1) allows to give the time to the stop method to check the if condition before the dispatcher thread terminates.

Reply:
If the `is_running` returns False, there is no need to join actually.

The main reason we join is so that the `run()` method will be fully stop when the next test runs.

Reply:
Yep. OK.